### PR TITLE
[runtime][HIP] Retire ROCm HAL backend

### PR DESCRIPTION
**Pull request by nithinsubbiah**
This patch retires the ROCm HAL backend given the new HIP HAL backend is functionally complete (https://github.com/openxla/iree/issues/15790). The new backend has been tested and proved to give on par or better performance compared to the ROCm backend.

The documentation still preserves the `ROCm` term to denote the AMD compute software while using `HIP` specifically for the runtime.

### CODE REVIEW

**Review by ScottTodd — COMMENTED** [review]
This isn't running CI because only `experimental/` and `docs/` files are changed. There are other places to update like
* https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/CMakeLists.txt#L583-L597
* https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/build_tools/pkgci/build_linux_packages.sh#L191
 
Once you update at least one of those then the CI will run.

File: docs/website/docs/guides/deployment-configurations/gpu-rocm.md
Comment (ScottTodd) [inline docs/website/docs/guides/deployment-configurations/gpu-rocm.md:48]: Thanks for the doc updates. I was also looking at these. Found a few other places to touch up: https://github.com/ScottTodd/iree/commit/c728bec69d2b820f99cec6e43502048fe1172c00

(I'm sort of waiting for a ROCM-capable dev machine so I can personally go through the setup instructions myself... it's hard to update docs with confidence when only working from remote machines that are already configured)

Comment (nithinsubbiah) [reply docs/website/docs/guides/deployment-configurations/gpu-rocm.md:48]: I just changed these main build commands so we're up to date if any user wants to build targeting AMD. If you're working on a comprehensive setup instructions overhaul, I can hold off on these changes

File: CMakeLists.txt
Comment (ScottTodd) [inline CMakeLists.txt:593]: https://github.com/openxla/iree/actions/runs/8665850617/job/23765392721?pr=17029#step:8:457
```
  CMake Error at runtime/src/iree/hal/drivers/CMakeLists.txt:40 (message):
    Driver passed to -DIREE_EXTERNAL_HAL_DRIVERS=rocm is not found.  Available:
    webgpu
```
That's what I was looking to see :P

Sorry, I also edited my earlier comment, the code to update to fix that error is here: https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/build_tools/pkgci/build_linux_packages.sh#L191

Comment (nithinsubbiah) [reply CMakeLists.txt:593]: ah I just pushed a fix for that, thanks for the catch!

File: docs/website/docs/guides/deployment-configurations/gpu-rocm.md
Comment (ScottTodd) [reply docs/website/docs/guides/deployment-configurations/gpu-rocm.md:48]: I'm not currently working on it. I started that commit then paused when I determined that it would be way easier with a local dev environment... which I might have in a few weeks. Help refreshing the docs would be great if you're up for it. Otherwise it'll be in my backlog.

Comment (nithinsubbiah) [reply docs/website/docs/guides/deployment-configurations/gpu-rocm.md:48]: Oh sure, I can take it up. Will do a follow up patch with (if any) updated instructions

**Comment by raikonenfnu** [comment]
+1 on Scott's comment on adding CI. Let's have PkgCI running on HIP before retiring experimental/ROCM-HAL backend.

https://github.com/openxla/iree/blob/main/experimental/regression_suite/tests/pregenerated/test_llama2.py

Here is my patch on adding ROCM to the PkgCI:
https://github.com/openxla/iree/commit/ebdb098b216c3e59a9977902823ede613f553f71

I think you just need to replace `device=rocm` to `device=hip`. :)

**Comment by ScottTodd** [comment]
> +1 on Scott's comment on adding CI. Let's have PkgCI running on HIP before retiring experimental/ROCM-HAL backend.
> 
> https://github.com/openxla/iree/blob/main/experimental/regression_suite/tests/pregenerated/test_llama2.py
> 
> Here is my patch on adding ROCM to the PkgCI: [ebdb098](https://github.com/openxla/iree/commit/ebdb098b216c3e59a9977902823ede613f553f71)
> 
> I think you just need to replace `device=rocm` to `device=hip`. :)

I enabled some ONNX tests using `--device=hip` recently in https://github.com/openxla/iree/pull/17014 . Should also update these: https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/experimental/regression_suite/tests/pregenerated/test_llama2.py#L226

**Comment by ScottTodd** [comment]
A few other files to update:
* https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/experimental/regression_suite/tests/pregenerated/test_ukernel.py#L159
* https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/samples/custom_dispatch/hip/kernels/example.mlir#L4

(regex search for `device.*rocm`)

**Review by stellaraccident — APPROVED** [review]
Woohoo. Thank you. Approving modulo other comments.

File: CMakeLists.txt
Comment (ScottTodd) [inline CMakeLists.txt:588]: Do we want to wait to merge this until issues like https://github.com/openxla/iree/issues/17033 are resolved? Maybe set a timeline on deprecation/removal of 1-2 weeks? We already have a few PRs in flight that touch this experimental code, so would be good to cut over ASAP and not need to maintain code scheduled for removal.

Comment (nithinsubbiah) [reply CMakeLists.txt:588]: Let's wait till next week to see if any more issues are reported. I'll take a look at the reported one and issue a PSA to use only HIP from now so we can stress test it further.

Comment (stellaraccident) [reply CMakeLists.txt:588]: Not doing on a Friday afternoon sgtm. But next week, I think we can afford to be aggressive on this. The plan and progress has been published for 6 months.

**Comment by ScottTodd** [comment]
Found another place to update that is not tested on CI (yet) but may affect some local development workflows: https://github.com/iree-org/iree/blob/568bb31780c3a04928b3cf62159e3272dd1465be/tests/e2e/matmul/BUILD.bazel#L544-L546

anything using that pattern should be updated to use the `hip` device too

File: docs/website/docs/guides/deployment-configurations/gpu-rocm.md
Comment (ScottTodd) [reply docs/website/docs/guides/deployment-configurations/gpu-rocm.md:48]: This file also has some references to the `rocm` HAL driver/device: https://github.com/iree-org/iree/blob/5aaaa79c30601985d12a0cdd081a6a3884ce55d8/docs/website/docs/guides/deployment-configurations/index.md?plain=1#L99

**Comment by ScottTodd** [comment]
> A few other files to update:
> 
> * https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/experimental/regression_suite/tests/pregenerated/test_ukernel.py#L159
> * https://github.com/openxla/iree/blob/40f25334d2b091a8805f7c52181ddcabd9924090/samples/custom_dispatch/hip/kernels/example.mlir#L4
> 
> (regex search for `device.*rocm`)

Bump. Those locations should still be updated.

**Review by ScottTodd — APPROVED** [review]
LGTM. Can you sync this to get past the merge conflicts?

**Comment by ScottTodd** [comment]
Can we please land this soon?

**Pull request merged**
